File: packages/java/src/main/java/ai/alumnium/Area.java
package ai.alumnium;

import ai.alumnium.Alumni.CheckOptions;
import ai.alumnium.Alumni.GetOptions;
import ai.alumnium.accessibility.BaseAccessibilityTree;
import ai.alumnium.client.Data;
import ai.alumnium.client.FindElementResult;
import ai.alumnium.client.HttpClient;
import ai.alumnium.client.HttpClient.ActionResult;
import ai.alumnium.client.HttpClient.PlanResult;
import ai.alumnium.driver.BaseDriver;
import ai.alumnium.result.DoResult;
import ai.alumnium.result.DoStep;
import ai.alumnium.tool.BaseTool;
import ai.alumnium.util.Retry;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class Area {
  private final int id;
  private final String description;
  private final BaseDriver driver;
  private final BaseAccessibilityTree accessibilityTree;
  private final Map<String, Class<? extends BaseTool>> tools;
  private final HttpClient client;

  public Area(
      int id,
      String description,
      BaseDriver driver,
      BaseAccessibilityTree accessibilityTree,
      Map<String, Class<? extends BaseTool>> tools,
      HttpClient client) {
    this.id = id;
    this.description = description;
    this.driver = driver;
    this.accessibilityTree = accessibilityTree;
    this.tools = tools;
    this.client = client;
  }

  public int id() {
    return id;
  }

  public String description() {
    return description;
  }

  public BaseDriver driver() {
    return driver;
  }

  public BaseAccessibilityTree accessibilityTree() {
    return accessibilityTree;
  }

  public Map<String, Class<? extends BaseTool>> tools() {
    return tools;
  }

  public HttpClient client() {
    return client;
  }

  /**
   * Act on the area.
   *
   * @param goal the goal to act on
   * @return the result of the action (explanation and executed steps)
   */
  public DoResult act(String goal) {
    return Retry.execute(
        () -> {
          driver.setAccessibilityTree(accessibilityTree);

          PlanResult response = client.planActions(goal, accessibilityTree.toStr(), driver.app());
          String explanation = response.explanation();
          List<String> steps = response.steps();
          List<DoStep> executedSteps = new ArrayList<>();
          for (String step : steps) {
            ActionResult actionResult =
                client.executeAction(goal, step, accessibilityTree.toStr(), driver.app());

            if (explanation.equals(goal)) {
              explanation = actionResult.explanation();
            }

            List<String> calledTools = new ArrayList<>();
            for (DoStep toolCall : actionResult.actions()) {
              calledTools.add(BaseTool.executeToolCall(toolCall, tools, driver));
            }
            executedSteps.add(new DoStep(step, calledTools));
          }
          return new DoResult(explanation, executedSteps);
        });
  }

  /**
   * Check a statement true or false within the area.
   *
   * @param statement the statement to check
   * @return the result of the check
   */
  public String check(String statement) {
    return check(statement, new CheckOptions(false));
  }

  /**
   * Check a statement true or false within the area.
   *
   * @param statement the statement to check
   * @param opts the options for the check
   * @return the result of the check
   */
  public String check(String statement, CheckOptions opts) {
    return Retry.execute(
        () -> {
          boolean vision = opts != null && opts.vision();
          HttpClient.RetrieveResult result =
              client.retrieve(
                  "Is the following true or false - " + statement,
                  accessibilityTree.toStr(),
                  driver.title(),
                  driver.url(),
                  vision ? driver.screenshot() : null,
                  driver.app());

          if (!Boolean.TRUE.equals(result.result().boxedValue())) {
            throw new AssertionError(result.explanation());
          }
          return result.explanation();
        });
  }

  /**
   * Get data from the area.
   *
   * @param data the data to get
   * @return the data
   */
  public Object get(String data) {
    return get(data, new GetOptions(false));
  }

  /**
   * Get data from the area.
   *
   * @param data the data to get
   * @param opts the options for the get
   * @return the data
   */
  public Object get(String data, GetOptions opts) {
    return Retry.execute(
        () -> {
          boolean vision = opts != null && opts.vision();
          HttpClient.RetrieveResult result =
              client.retrieve(
                  data,
                  accessibilityTree.toStr(),
                  driver.title(),
                  driver.url(),
                  vision ? driver.screenshot() : null,
                  driver.app());

          Data value = result.result();
          if (value == null || value.isNoop()) {
            return result.explanation();
          }
          return value.toObject();
        });
  }

  /**
   * Find an element in the area.
   *
   * @param description Natural language description of the element to find.
   * @return Native driver element (Selenium WebElement, Playwright Locator, or Appium WebElement).
   */
  public Object find(String description) {
    return Retry.execute(
        () -> {
          driver.setAccessibilityTree(accessibilityTree);
          FindElementResult response =
              client.findElement(description, accessibilityTree.toStr(), driver.app());
          return driver.findElement(response.id());
        });
  }
}

File: packages/python/src/alumnium/area.py
from retry import retry

from . import DELAY, RETRIES
from .accessibility.base_accessibility_tree import BaseAccessibilityTree
from .clients.http_client import HttpClient
from .clients.typecasting import Data
from .drivers import Element
from .drivers.base_driver import BaseDriver
from .logutils import get_logger
from .result import DoResult, DoStep
from .tools import BaseTool

logger = get_logger(__name__)


class Area:
    def __init__(
        self,
        id: int,
        description: str,
        driver: BaseDriver,
        accessibility_tree: BaseAccessibilityTree,
        tools: dict[str, BaseTool],
        client: HttpClient,
    ):
        self.id = id
        self.description = description
        self.driver = driver
        self.accessibility_tree = accessibility_tree
        self.tools = tools
        self.client = client

    @retry(tries=RETRIES, delay=DELAY, logger=logger)
    def do(self, goal: str) -> DoResult:
        """
        Executes a series of steps to achieve the given goal within the area.

        Args:
            goal: The goal to be achieved.

        Returns:
            DoResult containing the explanation and executed steps with their actions.
        """
        self.driver.set_accessibility_tree(self.accessibility_tree)

        explanation, steps = self.client.plan_actions(goal, self.accessibility_tree.to_str(), app=self.driver.app)

        executed_steps = []
        for step in steps:
            actor_explanation, actions = self.client.execute_action(
                goal, step, self.accessibility_tree.to_str(), app=self.driver.app
            )

            # When planner is off, explanation is just the goal — replace with actor's reasoning.
            if explanation == goal:
                explanation = actor_explanation

            called_tools = []
            for tool_call in actions:
                called_tool = BaseTool.execute_tool_call(tool_call, self.tools, self.driver)
                called_tools.append(called_tool)

            executed_steps.append(DoStep(name=step, tools=called_tools))

        return DoResult(explanation=explanation, steps=executed_steps)

    @retry(tries=RETRIES, delay=DELAY, logger=logger)
    def check(self, statement: str, vision: bool = False) -> str:
        """
        Checks a given statement true or false within the area.

        Args:
            statement: The statement to be checked.
            vision: A flag indicating whether to use a vision-based verification via a screenshot. Defaults to False.

        Returns:
            The summary of verification result.

        Raises:
            AssertionError: If the verification fails.
        """
        explanation, value = self.client.retrieve(
            f"Is the following true or false - {statement}",
            self.accessibility_tree.to_str(),
            title=self.driver.title,
            url=self.driver.url,
            screenshot=self.driver.screenshot if vision else None,
            app=self.driver.app,
        )
        assert value, explanation
        return explanation

    @retry(tries=RETRIES, delay=DELAY, logger=logger)
    def get(self, data: str, vision: bool = False) -> Data:
        """
        Extracts requested data from the area.

        Args:
            data: The data to extract.
            vision: A flag indicating whether to use a vision-based extraction via a screenshot. Defaults to False.

        Returns:
            The extracted data. If data cannot be extracted, returns the explanation string.
        """
        explanation, value = self.client.retrieve(
            data,
            self.accessibility_tree.to_str(),
            title=self.driver.title,
            url=self.driver.url,
            screenshot=self.driver.screenshot if vision else None,
            app=self.driver.app,
        )
        return explanation if value is None else value

    @retry(tries=RETRIES, delay=DELAY, logger=logger)
    def find(self, description: str) -> Element:
        """
        Finds an element within this area and returns the native driver element.

        Args:
            description: Natural language description of the element to find.

        Returns:
            Native driver element (Selenium WebElement, Playwright Locator, or Appium WebElement).
        """
        self.driver.set_accessibility_tree(self.accessibility_tree)
        response = self.client.find_element(description, self.accessibility_tree.to_str(), app=self.driver.app)
        return self.driver.find_element(response["id"])
